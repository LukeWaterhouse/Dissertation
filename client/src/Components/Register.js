import { useState, useContext } from 'react'
import axios from 'axios'
import UserContext from './UserContext'
import '../Css/Register.css'
import { useHistory } from 'react-router-dom'

function Register() {
  const [userName, setUserName] = useState('')
  const [password, setPassword] = useState('')
  const [userNameExists, setuserNameExists] = useState(false)

  const [isFieldBlank, setIsFieldBlank] = useState(false)

  const user = useContext(UserContext)
  let history = useHistory()

  function registerUser(e) {
    e.preventDefault()

    //sets error message if register fields are blank
    if (userName === '' || password == '') {
      setIsFieldBlank(true)
    } else {
      setIsFieldBlank(false)

      //sends request to backend to store new user if they don't exist
      const data = { userName, password }
      axios
        .post('http://localhost:5000/register', data, { withCredentials: true })
        .then((response) => {
          console.log(response.data)
          if (response.data === 'userNameExists') {
            setuserNameExists(true)
          } else {
            console.log('Posted!')
            user.setUserName(response.data.userName)
            setUserName('')
            setPassword('')
            setuserNameExists(false)
            history.push('/homepage')
          }
        })
    }
  }

  return (
    <div className="container">
      <h3>Welcome to Data House</h3>
      <h4>Register</h4>
      <form action="" onSubmit={(e) => registerUser(e)}>
        {isFieldBlank && (
          <div style={{ color: 'red' }}>Please fill out both fields!</div>
        )}
        {userNameExists && (
          <div style={{ color: 'red' }}>Username already exists!</div>
        )}

        <div style={{ color: 'white' }}>Username: </div>
        <input
          type="text"
          placeholder="Username"
          value={userName}
          maxLength="15"
          onChange={(e) => setUserName(e.target.value)}
        />
        <br />

        <div style={{ color: 'white', paddingTop: '10px' }}>Password</div>
        <input
          type="password"
          placeholder="password"
          value={password}
          maxLength="20"
          onChange={(e) => setPassword(e.target.value)}
        />
        <br />
        <div
          style={{
            paddingTop: '20px',
            display: 'flex',
            justifyContent: 'center'
          }}
        >
          <button className="btn-1" type="submit">
            REGISTER
          </button>
        </div>
      </form>
    </div>
  )
}

export default Register
