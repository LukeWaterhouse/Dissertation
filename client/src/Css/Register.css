* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}

body {
  display: flex;
  height: 100vh;
  justify-content: center;
  align-items: center;
}

.container {
  max-width: 700px;
  width: 100%;
  background: #4e6839;
  padding: 20px;
}

h3 {
  color: white;
  font-size: xx-large;
  font-weight: 400;
  padding-bottom: 50px;
}

h4 {
  color: white;
  font-size: x-large;
  font-weight: 200;
  padding-bottom: 10px;
}

.btn-1 {
  font-size: 15px;
  background: #354528;
  border-radius: 0;
  padding: 10px;
  border-style: none;
  color: rgb(233, 233, 233);
}

.btn-1:hover {
  background: #415630;
  transition: 0.4s;
}
