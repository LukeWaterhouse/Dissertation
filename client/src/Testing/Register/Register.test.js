import React from 'react'
import ReactDOM from 'react-dom'
import { render, cleanup, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import Register from '../../Components/Register'

afterEach(cleanup)

it('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(<Register></Register>, div)
  screen.debug
})

it('error message shows if you click register without entering info', () => {
  render(<Register />)
  userEvent.click(screen.getByRole('button'))
  expect(screen.getByText(/Please fill out both fields!/i)).toBeInTheDocument()
})

it('error message shows if you click register without entering info', () => {
  render(<Register />)
  userEvent.type(screen.getByRole('textbox'), 'sampleUsername')
  userEvent.type(screen.getByPlaceholderText(/password/i), 'samplePassword')
  userEvent.click(screen.getByRole('button'))

  expect(screen.queryByText('Please fill out both fields!')).toBeNull()
})

it('register matches snapshot', () => {
  const tree = renderer.create(<Register />).toJSON()
  expect(tree).toMatchSnapshot()
})
