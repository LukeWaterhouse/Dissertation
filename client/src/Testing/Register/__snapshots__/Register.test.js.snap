// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`register matches snapshot 1`] = `
<div
  className="container"
>
  <h3>
    Welcome to GeckoSpot
  </h3>
  <h4>
    Register
  </h4>
  <form
    action=""
    onSubmit={[Function]}
  >
    <div
      style={
        Object {
          "color": "white",
        }
      }
    >
      Username: 
    </div>
    <input
      maxLength="15"
      onChange={[Function]}
      placeholder="Username"
      type="text"
      value=""
    />
    <br />
    <div
      style={
        Object {
          "color": "white",
          "paddingTop": "10px",
        }
      }
    >
      Password
    </div>
    <input
      maxLength="20"
      onChange={[Function]}
      placeholder="password"
      type="password"
      value=""
    />
    <br />
    <div
      style={
        Object {
          "display": "flex",
          "justifyContent": "center",
          "paddingTop": "20px",
        }
      }
    >
      <button
        className="btn-1"
        type="submit"
      >
        REGISTER
      </button>
    </div>
  </form>
</div>
`;
